Src/LiquidProjections/Dispatcher.cs

```diff
@@ -10,62 +10,92 @@ public class Dispatcher
+            {
+                HandleTransactions = async (transactions, info) => await HandleTransactions(transactions, handler, info),
+                NoSuchCheckpoint = async info => await HandleUnknownCheckpoint(info, handler, options)
+            }, options.Id);
+        }
 
-            lock (subscriptionMonitor)
+        private static async Task HandleTransactions(IReadOnlyList<Transaction> transactions, Func<IReadOnlyList<Transaction>, SubscriptionInfo, Task> handler, SubscriptionInfo info)
+        {
+            try
             {
-                subscription = createSubscription(checkpoint,
-                    async transactions =>
-                    {
-                        try
-                        {
-                            await handler(transactions);
-                        }
-                        catch (Exception exception)
-                        {
-                            LogProvider.GetLogger(typeof(Dispatcher)).FatalException(
-                                "Projector exception was not handled. Event subscription has been cancelled.",
-                                exception);
+                await handler(transactions, info);
+            }
+            catch (Exception exception)
+            {
+                LogProvider.GetLogger(typeof(Dispatcher)).FatalException(
+                    "Projector exception was not handled. Event subscription has been cancelled.",
+                    exception);
 
-                            lock (subscriptionMonitor)
-                            {
-                                subscription.Dispose();
-                            }
-                        }
-                    },
-                    subscriptionId);
+                info.Subscription?.Dispose();
             }
+        }
+
+        private async Task HandleUnknownCheckpoint(SubscriptionInfo info, Func<IReadOnlyList<Transaction>, SubscriptionInfo, Task> handler, SubscriptionOptions options)
+        {
+            if (options.RestartWhenAhead)
+            {
+                try
+                {
+                    info.Subscription?.Dispose();
+
+                    await options.BeforeRestarting();
 
-            return subscription;
+                    Subscribe(null, handler, options);
+                }
+                catch (Exception exception)
+                {
+                    LogProvider.GetLogger(typeof(Dispatcher)).FatalException(
+                        "Failed to restart the projector.",
+                        exception);
+                }
+            }
         }
+    }
 
-        public IDisposable Subscribe(long? checkpoint, Func<IReadOnlyList<Transaction>, Task> handler)
+    public class SubscriptionOptions
+    {
+        public SubscriptionOptions()
         {
-            return Subscribe(checkpoint, handler, null);
+            BeforeRestarting = () => Task.FromResult(0);

```

Comment:
Syntax suggestion: you can assign the initial value in the property declaration:
`public Func<Task> BeforeRestarting { get; set; } = () => Task.FromResult(0);`